// Copyright 2023 The Centipede Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Proto representation of Centipede coverage reports.
syntax = "proto3";

package centipede;

// Describes the covered edges and functions from, e.g., a fuzzing run.
//
// Next tag: 2
message CoverageReport {
  // Describes a particular location within a program.
  //
  // Next tag: 5
  message Edge {
    // The name of the function where the edge resides.
    string function_name = 1;
    // The name of the file where the edge resides.
    string file_name = 2;
    // The line in `file_name` where the edge resides.
    int32 line = 3;
    // The column in `file_name` where the edge resides.
    int32 column = 4;
  }
  repeated Edge covered_edges = 1;
}

// Describes the comparison of the coverage of two corpora A and B (--analyze).
// Next tag: 5
message AnalyzeReport {
  // Coverage from corpus A.
  CoverageReport a_coverage = 1;
  // Coverage from corpus B.
  CoverageReport b_coverage = 2;
  // Coverage found in A but not B
  CoverageReport a_only_coverage = 3;
  // Coverage found in B but not A
  CoverageReport b_only_coverage = 4;
}
